File: Sources/HoleyShapes/HoleyShape.swift
/// Every shape included in the original `holey-shapes` catalog.
public enum HoleyShape: String, CaseIterable, Identifiable, Sendable {
  case disc
  case roundBlock = "round-block"
  case hex
  case capsule
  case prism
  case cross
  case triangle
  case diamond
  case sunburst
  case octagon
  case chevron
  case longBar = "long-bar"
  case flowerStar = "flower-star"
  case flower
  case bowtie

  public var id: String { rawValue }

  public var name: String {
    switch self {
    case .disc: "Perforated disc"
    case .roundBlock: "Round block"
    case .hex: "Hex slab"
    case .capsule: "Capsule"
    case .prism: "Cut prism"
    case .cross: "Cross block"
    case .triangle: "Triangle plate"
    case .diamond: "Diamond tile"
    case .sunburst: "Sunburst slab"
    case .octagon: "Octagon"
    case .chevron: "Bent chevron"
    case .longBar: "Long bar"
    case .flowerStar: "Five-point bloom"
    case .flower: "Daisy flower"
    case .bowtie: "Bowtie slab"
    }
  }

  public var color: HoleyColor {
    switch self {
    case .disc: HoleyColor(red: 0x63, green: 0x37, blue: 0xFF)
    case .roundBlock: HoleyColor(red: 0x9B, green: 0xED, blue: 0x00)
    case .hex: HoleyColor(red: 0xFF, green: 0x08, blue: 0x78)
    case .capsule: HoleyColor(red: 0x28, green: 0x78, blue: 0xFF)
    case .prism: HoleyColor(red: 0xFF, green: 0x6A, blue: 0x00)
    case .cross: HoleyColor(red: 0x00, green: 0xDB, blue: 0xFF)
    case .triangle: HoleyColor(red: 0x00, green: 0xD6, blue: 0x9F)
    case .diamond: HoleyColor(red: 0xFF, green: 0xD0, blue: 0x00)
    case .sunburst: HoleyColor(red: 0xB9, green: 0x4C, blue: 0xFF)
    case .octagon: HoleyColor(red: 0xFF, green: 0x38, blue: 0x38)
    case .chevron: HoleyColor(red: 0x00, green: 0xC9, blue: 0xF2)
    case .longBar: HoleyColor(red: 0xFF, green: 0x4F, blue: 0xA3)
    case .flowerStar: HoleyColor(red: 0xFF, green: 0x76, blue: 0x57)
    case .flower: HoleyColor(red: 0xC9, green: 0x4D, blue: 0xFF)
    case .bowtie: HoleyColor(red: 0x00, green: 0xD7, blue: 0xB9)
    }
  }

  public var backgroundColor: HoleyColor {
    switch self {
    case .disc: HoleyColor(red: 0xD7, green: 0xCE, blue: 0xFF)
    case .roundBlock: HoleyColor(red: 0xE7, green: 0xFF, blue: 0x92)
    case .hex: HoleyColor(red: 0xFF, green: 0xD6, blue: 0xE4)
    case .capsule: HoleyColor(red: 0xCC, green: 0xED, blue: 0xFF)
    case .prism: HoleyColor(red: 0xFF, green: 0xD4, blue: 0xA3)
    case .cross: HoleyColor(red: 0xD8, green: 0xD5, blue: 0xCF)
    case .triangle: HoleyColor(red: 0xB9, green: 0xF4, blue: 0xDF)
    case .diamond: HoleyColor(red: 0xFF, green: 0xF0, blue: 0xA6)
    case .sunburst: HoleyColor(red: 0xE6, green: 0xD4, blue: 0xFA)
    case .octagon: HoleyColor(red: 0xFF, green: 0xD0, blue: 0xC8)
    case .chevron: HoleyColor(red: 0xC9, green: 0xF3, blue: 0xFF)
    case .longBar: HoleyColor(red: 0xFF, green: 0xE0, blue: 0xEE)
    case .flowerStar: HoleyColor(red: 0xFF, green: 0xE1, blue: 0xD8)
    case .flower: HoleyColor(red: 0xF2, green: 0xD8, blue: 0xFF)
    case .bowtie: HoleyColor(red: 0xD0, green: 0xF7, blue: 0xF0)
    }
  }

  public var defaultHoleCount: Int {
    switch self {
    case .disc, .hex, .cross, .diamond, .sunburst, .longBar, .flowerStar: 5
    case .roundBlock, .capsule, .prism, .triangle, .chevron: 4
    case .octagon, .flower, .bowtie: 6
    }
  }

  public var maximumHoleCount: Int {
    8
  }
}

public struct HoleyShapeInfo: Hashable, Identifiable, Sendable {
  public let shape: HoleyShape
  public let name: String
  public let color: HoleyColor

  public var id: String { shape.rawValue }
}

public enum HoleyShapes {
  public static let catalog: [HoleyShapeInfo] = HoleyShape.allCases.map {
    HoleyShapeInfo(shape: $0, name: $0.name, color: $0.color)
  }
}

File: src/main/kotlin/com/aakashreddy/holeyshapes/HoleyShape.kt
package com.aakashreddy.holeyshapes

import androidx.compose.runtime.Immutable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import kotlin.math.roundToInt

/** Every shape included in the original holey-shapes catalog. */
enum class HoleyShape(
    val key: String,
    val displayName: String,
    val color: Color,
    val backgroundColor: Color,
    val defaultHoleCount: Int,
) {
    Disc("disc", "Perforated disc", Color(0xFF6337FF), Color(0xFFD7CEFF), 5),
    RoundBlock("round-block", "Round block", Color(0xFF9BED00), Color(0xFFE7FF92), 4),
    Hex("hex", "Hex slab", Color(0xFFFF0878), Color(0xFFFFD6E4), 5),
    Capsule("capsule", "Capsule", Color(0xFF2878FF), Color(0xFFCCEDFF), 4),
    Prism("prism", "Cut prism", Color(0xFFFF6A00), Color(0xFFFFD4A3), 4),
    Cross("cross", "Cross block", Color(0xFF00DBFF), Color(0xFFD8D5CF), 5),
    Triangle("triangle", "Triangle plate", Color(0xFF00D69F), Color(0xFFB9F4DF), 4),
    Diamond("diamond", "Diamond tile", Color(0xFFFFD000), Color(0xFFFFF0A6), 5),
    Sunburst("sunburst", "Sunburst slab", Color(0xFFB94CFF), Color(0xFFE6D4FA), 5),
    Octagon("octagon", "Octagon", Color(0xFFFF3838), Color(0xFFFFD0C8), 6),
    Chevron("chevron", "Bent chevron", Color(0xFF00C9F2), Color(0xFFC9F3FF), 4),
    LongBar("long-bar", "Long bar", Color(0xFFFF4FA3), Color(0xFFFFE0EE), 5),
    FlowerStar("flower-star", "Five-point bloom", Color(0xFFFF7657), Color(0xFFFFE1D8), 5),
    Flower("flower", "Daisy flower", Color(0xFFC94DFF), Color(0xFFF2D8FF), 6),
    Bowtie("bowtie", "Bowtie slab", Color(0xFF00D7B9), Color(0xFFD0F7F0), 6),
    ;

    val maximumHoleCount: Int get() = 8

    companion object {
        fun fromKey(key: String): HoleyShape? = entries.firstOrNull { it.key == key }
    }
}

/** Rendering options for [HoleyShapeView]. The renderer clamps numeric values to supported ranges. */
@Immutable
data class HoleyShapeConfig(
    val shape: HoleyShape = HoleyShape.Disc,
    val faceColor: Color = shape.color,
    val shadowColor: Color = faceColor.darkened(),
    val holes: Int = shape.defaultHoleCount,
    val seed: Int = 0,
    val animated: Boolean = true,
    val durationMillis: Int = 900,
    val shadowX: Float = 18f,
    val shadowY: Float = 21f,
    val shadowSteps: Int = 12,
) {
    internal fun normalized() = copy(
        holes = holes.coerceIn(0, shape.maximumHoleCount),
        seed = seed.coerceIn(-1_000_000, 1_000_000),
        durationMillis = durationMillis.coerceIn(100, 10_000),
        shadowX = shadowX.coerceIn(-80f, 80f),
        shadowY = shadowY.coerceIn(-80f, 80f),
        shadowSteps = shadowSteps.coerceIn(1, 32),
    )

    fun shuffled(): HoleyShapeConfig = copy(
        seed = if (seed >= 1_000_000) -1_000_000 else seed + 1,
    )

    companion object {
        /** Creates a stable shape and hole layout for an identity such as a user ID. */
        fun seeded(identity: String, animated: Boolean = true): HoleyShapeConfig {
            var hash = -3_750_763_034_362_895_579L
            identity.encodeToByteArray().forEach { byte ->
                hash = hash xor (byte.toLong() and 0xFF)
                hash *= 1_099_511_628_211L
            }
            val random = SplitMix64(hash)
            val shape = HoleyShape.entries[(random.next().toULong() % HoleyShape.entries.size.toUInt()).toInt()]
            val seed = (random.next().toULong() % 1_000_001u).toInt()
            return HoleyShapeConfig(shape = shape, seed = seed, animated = animated)
        }
    }
}

/** Catalog metadata in the same order as the web and Swift packages. */
object HoleyShapes {
    val catalog: List<HoleyShape> = HoleyShape.entries
}

internal fun Color.darkened(factor: Float = 0.52f): Color {
    val argb = toArgb()
    fun channel(shift: Int): Int {
        val value = (argb shr shift) and 0xFF
        return (value * factor).roundToInt().coerceIn(0, 255)
    }
    return Color(
        red = channel(16),
        green = channel(8),
        blue = channel(0),
        alpha = (argb ushr 24) and 0xFF,
    )
}

private class SplitMix64(private var state: Long) {
    fun next(): Long {
        state += -7_046_029_254_386_353_131L
        var value = state
        value = (value xor (value ushr 30)) * -4_658_895_280_553_007_687L
        value = (value xor (value ushr 27)) * -7_723_592_293_110_705_685L
        return value xor (value ushr 31)
    }
}
